base/shared/src/main/scala/scalaz/std/utils.scala

```diff
@@ -1,12 +1,17 @@
 package scalaz
 package std
 
-import core.EqClass
+import scala.{ AnyRef, AnyVal }
+
+import core.{ EqAnyRef, EqClass }
 import data.Cord
 import debug.DebugClass
 
 private[std] object utils {
-  def singletonEq[A]: Eq[A]      = instanceOf[EqClass[A]]((a, b) => true)
-  def universalEq[A]: Eq[A]      = instanceOf[EqClass[A]]((a, b) => a == b)
+  def singletonEq[A]: Eq[A] = instanceOf(((a: A, b: A) => true): EqClass[A])
+
+  def universalEqAnyRef[A <: AnyRef]: Eq[A] = instanceOf(((a: A, b: A) => a == b): EqAnyRef[A])
+  def universalEqAnyVal[A <: AnyVal]: Eq[A] = instanceOf(((a, b) => a == b): EqClass[A])
```

Comment:
if we're going for speed I'd say this should not be done: this will go through `BoxesRunTime` to handle the case in which `a` and `b` aren't the same type (whereas we know they will be). Writing out `instanceOf(_ == _)` for every type where it makes sense should allow the compiler to do a less complicated equality comparison.